crypto/evp/e_chacha20_poly1305.c

```diff
@@ -193,23 +195,158 @@ static int chacha20_poly1305_init_key(EVP_CIPHER_CTX *ctx,
     return 1;
 }
 
+#  if !defined(OPENSSL_SMALL_FOOTPRINT)
+static const unsigned char zero[2 * CHACHA_BLK_SIZE] = { 0 };
```

Comment:
Side-note: The initializer isn't needed.
Not a side-note: should this be in the function?

Reply:
> Side-note: The initializer isn't needed.

I know. It's readability thing. Being explicit spares brain power recalling what does no initializer means.

> Not a side-note: should this be in the function?

It's used in both functions, new and old. Moving it into function would make old one declare own. It would be waste.

Reply:
> Side-note: The initializer isn't needed.

On additional counter note one can observe that without initializer gcc compiles it as .comm, e.g. "common" variable with no section "affiliation". As result it ends up in .bss, writable section. In other words omission of initializer effectively renders const ineffective. In sense that customarily you'd expect to find const variables in .rodata section. As additional insurance that if pointers are corrupted, attempt to modify variable would trigger SEGV.

Reply:
Interesting. That looks like a compiler bug (or maybe rather a missing feature).

Reply:
Well, formally speaking standard doesn't obliges compiler to put const data into .rodata. It rather prohibits generating code that would "consciously" modify const data, that's all. It just so happens that programmers who are accustomed to contemporary VM benefits **expect** const data to be placed into .rodata and appreciate it. One should also keep in mind that it's not always possible to put const data into .rodata. Consider `const char * const foo = "bar";` Compile without and with -fPIC. Trouble is that with -fPIC foo can't be resolved at link time and hence can't be placed into .rodata. Now, contemporary Linux would actually collect relocations into separate segment and have run-time linker protect that, but this is not universal thing. In sense that you can find OSes and even compilers for Linux that would place foo into writable segment.